mrmustard/lab_dev/circuit_components.py

```diff
@@ -233,6 +233,34 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):
+            # self has ket and bra, other has ket or bra
+            return self @ other @ other.adjoint
+
+        if (self.wires.ket or self.wires.bra) and other.wires.ket and other.wires.bra:
```

Comment:
now that i think of it: isn't `self.wires.ket or self.wires.bra` always `True`?

Reply:
Oh I think you are right. It would be one of them is True but not both of them...Emmmm needs a better condition here

Reply:
We could, but that's already taken care of by the matmul, when it does ``self.wires @ other.wires`` to find the wires of the returned object. If the two objects have invalid input/output wires (for example, ``vacuum >> vacuum``), it will error

In other words, the error belongs to the lower level implementation